pennylane/workflow/qnode.py

```diff
@@ -851,8 +896,8 @@ def _impl_call(self, *args, **kwargs) -> Result:
 
         return _to_qfunc_output_type(res, self._qfunc_output, tape.shots.has_partitioned_shots)
 
-    def __call__(self, *args, **kwargs) -> Result:
-        if "shots" in kwargs and qml.set_shots in self.transform_program:
+    def __call__(self, *args, **kwargs) -> qml.typing.Result:
+        if "shots" in kwargs and self._shots_override_device:
```

Comment:
Actually, can we just move this warning up to where we are deciding shots in the start of `construct`? I think it would make things more clear to have all the shots logic in one place.

Reply:
True